crates/weaver_semconv/src/body.rs

```diff
@@ -0,0 +1,361 @@
+    /// The collection of body fields associated with a body definition
+    Fields {
+        /// Identifies that the type of the body is a map of fields or a string.
+        r#type: BodyType,
+        /// A brief description of the body.
+        #[serde(skip_serializing_if = "String::is_empty")]
+        #[serde(default)]
+        brief: String,
+        /// A more elaborate description of the body.
+        /// It defaults to an empty string.
+        #[serde(skip_serializing_if = "String::is_empty")]
+        #[serde(default)]
+        note: String,
+        /// Specifies the stability of the body.
+        #[serde(skip_serializing_if = "Option::is_none")]
+        stability: Option<Stability>,
+        /// Sequence of example values for the body or single example
+        /// value. They are required only for string types. Example values
+        /// must be of the same type of the body. If only a single example is
+        /// provided, it can directly be reported without encapsulating it
+        /// into a sequence/dictionary.
+        #[serde(skip_serializing_if = "Option::is_none")]
+        examples: Option<Examples>,
+        /// Identifies the definition of the "fields" of the body when the body type is "map".
+        #[serde(skip_serializing_if = "Vec::is_empty")]
+        fields: Vec<BodyFieldSpec>,
+    },
+    /// The body will just be a string.
+    String {
+        /// Identifies that the type of the body is a string.
+        r#type: BodyType,
+        /// A brief description of the body.
+        #[serde(skip_serializing_if = "String::is_empty")]
+        #[serde(default)]
+        brief: String,
+        /// A more elaborate description of the body.
+        /// It defaults to an empty string.
+        #[serde(skip_serializing_if = "String::is_empty")]
+        #[serde(default)]
+        note: String,
+        /// Specifies the stability of the body.
+        #[serde(skip_serializing_if = "Option::is_none")]
+        stability: Option<Stability>,
+        /// Sequence of example values for the body or single example
+        /// value. They are required only for string types. Example values
+        /// must be of the same type of the body. If only a single example is
+        /// provided, it can directly be reported without encapsulating it
+        /// into a sequence/dictionary.
+        #[serde(skip_serializing_if = "Option::is_none")]
+        examples: Option<Examples>,
+    },
+}
+
+impl BodySpec {
+    /// Returns true if the body field is required.
+    #[must_use]
+    pub fn has_fields(&self) -> bool {
+        match self {
+            BodySpec::Fields { fields, .. } => !fields.is_empty(),
+            BodySpec::String { .. } => false,
+        }
+    }
+}
+
+/// Identifies the different types of body (specification).
+#[derive(Serialize, Deserialize, Debug, Clone, Eq, PartialEq, Hash, JsonSchema)]
+#[serde(rename_all = "snake_case")]
+pub enum BodyType {
+    /// A map body type.
+    Map,
```

Comment:
I think it'd be better if this enum actually had values.  E.g.  Map would have nested `TreeMap<String, BodyType>` or `TreeMap<String,BodyField>` -> where `BodyField` has `requirement_level`, `notes`, etc.

I don't think my point about not having two ways of representing compositionality made it through.   What I want to avoid here is having TWO enums you need to walk through to understand the type system in semconv.   Right now, `Map` doesn't really cover the fields in the map, meaning I need to walk through two strutures.

I don't have a ton of free time today, but I can put together an example/suggested hierarchy here, but I envision something like:


```rust
trait BodySpec {
  // Whatever serde flag denotes tagged-disjoin union
  r#type: BodyType
}
enum BodyType {
  String(CommonAttributes),
  Map {
     fields: TreeMap<String, BodyType>,
     // whatever serde attribute denotes flattening.
     attributes: CommonAttributes,
  }
}
struct CommonAttributes {
  requirement_level: RequirementLevel,
  notes: Option<String>,
  ...
}
```

Reply:
OK, I think I'm following your suggestion and fundamentally it ends up representing the same structured output (without the existing duplication), not sure I'll get to (and complete) today before tomorrows meeting but I think I like this direction (I guess I'm still getting my head wrapped around the rust structured enum definition / usage)

Reply:
Created draft PR with some simple examples (tests) to discuss the definition, I'm going to start to merge the changes from that PR into this one (and any changes we make to that one)
https://github.com/open-telemetry/weaver/pull/365